Realm/RLMQueryUtil.mm

```diff
@@ -190,64 +201,80 @@ bool has_any_to_many_links() const {
         });
     }
 
-    util::Optional<ColumnReference> last_link_column() const {
-        if (!m_links.size()) {
-            return util::none;
-        }
-        auto links = m_links;
-        RLMProperty *property = links.back();
-        links.resize(links.size() - 1);
-        return ColumnReference(m_schema, property, std::move(links));
+    ColumnReference last_link_column() const {
+        REALM_ASSERT(!m_links.empty());
+        return {*m_query, *m_group, m_schema, m_links.back(), {m_links.begin(), m_links.end() - 1}};
     }
 
-    ColumnReference column_ignoring_links() const {
-        return {m_schema, m_property};
+    ColumnReference column_ignoring_links(Query& query) const {
+        return {query, *m_group, m_schema, m_property};
     }
 
 private:
-    Table* table_for_query(Query& query) const
+    template <typename T, typename... SubQuery>
+    auto resolve_backlink(std::true_type, SubQuery&&... subquery) const
```

Comment:
What do you think of the following:

    struct ConstexprIf {
        template<typename Then, typename Else>
        static auto call(std::true_type, Then&& fn, Else&&) {
            return fn();
        }

        template<typename Then, typename Else>
        static auto call(std::false_type, Then&&, Else&& fn) {
            return fn();
        }
    };

    class ColumnReference {
        template <typename T, typename... SubQuery>
        auto resolve(SubQuery&&... subquery) const
        {
            static_assert(sizeof...(SubQuery) < 2, "resolve() takes at most one subquery");
            set_link_chain_on_table();
            if (type() != RLMPropertyTypeLinkingObjects) {
                return m_table->template column<T>(index(), std::move(subquery)...);
            }
            else {
                return ConstexprIf::call(std::is_same<T, Link>(), [&](auto...) {
                    return with_link_origin(m_property, [&](Table& table, size_t col) {
                        return m_table->template column<T>(table, col, std::move(subquery)...);
                    });
                }, [](auto...) -> Columns<T> { __builtin_unreachable(); });
            }
        }
    };

This fixes the locality issue and makes it much more clear what the true_type/false_type is doing (and that it's faking `if constexpr` via tag dispatch), but there's also some dreadfully subtle bits such as that the `auto...` is there to prevent the body of the lambda from being instantiated when the other case is used.